### PR TITLE
Add timeout to gRPC requests

### PR DESCRIPTION
**Pull request by ghost**
This PR introduces a backoff module that handles function calls with configurable timeout, which also increases with every timed-out call. This is then used for all non-streaming gRPC-related methods of the Client. The default timeout is set to 300 seconds (i.e. 5min), with flexibility to adjust as needed.

### CODE REVIEW

**Comment by ghost** [comment]
This timeout is now increased dynamically with each failed connection attempt.

**Review by tiyash-basu-frequenz — COMMENTED** [review]
It looks like the `grpc_call_with_timeout` is invoked per call. Maybe I am wrong, but wasn't the idea to implement the backoff at the client level, and not at a call level?

If we indeed need to implement it at a client level, then we could send the commands via a channel to a central client task, that can impose a global backoff logic. It would also need logic to reset the backoff after a certain period passes since the last successful RPC call.

**Comment by ghost** [comment]
Hmm, fair point that the timeout could be increased at the client level. That said, some gRPCs inherently take much longer than others (e.g., list_gridpool_orders), which is why I was thinking of having this handled at the gRPC level. However, I’m definitely not opposed to managing it at the client level either.

For me, the most important thing is that we have a timeout in place (whether it’s dynamically decided or fixed) to avoid hanging calls.

**Comment by tiyash-basu-frequenz** [comment]
I see. Different RPCs having different timeouts needs to be considered, true.

The main objective of timeouts would be to reduce stress on the client's connection. Having per-call timeout violates this principle. One call might be backing off, but then a new call might go straight through. So my guess is that while you may see some benefits of this approach right away, it would be a matter of chance when the connection to the server gets stressed. The more client apps, the further it will regress.

**Comment by ghost** [comment]
I currently have a `max_timeout` (maximum timeout duration for gRPC calls), `timeout_increment` (increment value for timeout on each retry, e.g. 20secs) and a `max_timeout_retries` (maximum number of retry attempts when a timeout is reached). Having those 3 vars could ba a bit of an overkill, I could remove one of them if needed.

File: src/frequenz/client/electricity_trading/_client.py [hunk 1]
```diff
@@ -116,15 +121,28 @@ def __new__(
...
     ) -> None:
         """Initialize the client.
 
         Args:
             server_url: The URL of the Electricity Trading service.
             connect: Whether to connect to the server immediately.
             auth_key: The API key for the authorization.
+            initial_timeout: Initial timeout duration for gRPC calls (in seconds).
+            max_timeout: Maximum timeout duration for gRPC calls (in seconds).
+            timeout_increment: Increment value for timeout on each retry (in seconds).
```
Comment (tiyash-basu-frequenz) [inline src/frequenz/client/electricity_trading/_client.py]: I would suggest making this a timeout multiplier instead. Instead of adding a value, you would then multiply it with this factor. That way you can achieve exponential backing off easily.

File: src/frequenz/client/electricity_trading/_client.py [hunk 2]
```diff
@@ -159,6 +177,87 @@ def __init__(
...
+                self._initial_timeout,
+                self._reset_interval,
+            )
+
+    async def grpc_call_with_timeout(
+        self,
+        call: Callable[..., Awaitable[Any]],
+        *args: Any,
+        **kwargs: Any,
+    ) -> Any:
```
Comment (tiyash-basu-frequenz) [inline src/frequenz/client/electricity_trading/_client.py]: I would suggest extracting the timeout logic to a separate `backoff` module, so that you can unit test it. You could just use that `backoff` module here.

**Review by tiyash-basu-frequenz — COMMENTED** [review]
The addition itself looks good. I have a few improvement suggestions.

File: src/frequenz/client/electricity_trading/_client.py (hunk 1 above)
Comment (ghost) [reply src/frequenz/client/electricity_trading/_client.py]: Oh okay I actually purposefully did an increment and not multiplier as I thought having an exponential increase would be too much here. But I can change it to multiplier for sure if it's better practice!

File: src/frequenz/client/electricity_trading/_client.py (hunk 2 above)
Comment (ghost) [reply src/frequenz/client/electricity_trading/_client.py]: Done

File: src/frequenz/client/electricity_trading/_backoff.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
+# License: MIT
+# Copyright © 2024 Frequenz Energy-as-a-Service GmbH
+
+"""Module to define a backoff strategy to handle non-successful gRPC connection attempts."""
```
Comment (tiyash-basu-frequenz) [inline src/frequenz/client/electricity_trading/_backoff.py]: Nit - What you have achieved here is a generic backoff module, which can be used for use-caes besides gRPC as well. So this line needs to be updated :)

File: src/frequenz/client/electricity_trading/_backoff.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+import asyncio
+import logging
+from datetime import datetime, timezone
+from typing import Any, Awaitable, Callable
+
+_logger = logging.getLogger(__name__)
+
+
+class Backoff:
```
Comment (tiyash-basu-frequenz) [inline src/frequenz/client/electricity_trading/_backoff.py]: Can you add unit tests for this module please? At least the following cases should be covered:

1. success on attempt 1
2. success on an attempt > 1 but before reaching max timeout
3. success on attempt > 1 and after reaching max timeout for 2 iterations
4. failure
5. backoff reset
6. increment

**Review by tiyash-basu-frequenz — COMMENTED** [review]
Looks very good! I would just suggest adding a few tests to nail down stability and prevent regression, and done.

File: src/frequenz/client/electricity_trading/_backoff.py (hunk 4 above)
Comment (ghost) [reply src/frequenz/client/electricity_trading/_backoff.py]: Yess on it, I started yesterday but not fully done yet. Will finish them now!

Comment (ghost) [reply src/frequenz/client/electricity_trading/_backoff.py]: Some other topics came up yesterday but just finished them up now!:)

**Comment by shsms** [comment]
Sorry, I'm just realizing this has become a retry implementation for non-streaming calls.

Retries of this sort have to be infinite because we have to keep trying until the API call succeeds, and doing that in the client would take away the actor's agency in deciding what to do when there is a client error, like taking fallback action, sending a slack message, etc.

It also introduces the risk of creating orders with outdated power or price forecasts if the client's retry succeeds after a few minutes.

For this reason, all of our clients should have retries *only* for streaming APIs (which are taken care of by the `GrpcStreamBroadcaster` in the base client).

For non-streaming APIs, retries can be implemented in the actors using the `ExponentialBackoff` provided by the base client.  I'm also adding a receiver interface for this in the base client, allowing us to use a `select` loop to decide when to retry.

Also, I've implemented these in the actor for now in this PR: https://github.com/frequenz-io/frequenz-actor-electricity-trading/pull/249

So, I think this PR should go back to just implementing a timeout for non-streaming methods to prevent them from blocking for too long and limiting the actor's ability to take fallback action.

**Comment by tiyash-basu-frequenz** [comment]
Then all that needs to be done is invoke only simple RPCs with the backoff, and call the streaming ones directly. If a backoff module exists in the base client, then that could be used directly (and sorry that you implemented it here @camille-bouvy-frequenz).

I also would not recommend going back to a non-global backoff, since I do not see a value in that.

**Comment by shsms** [comment]
> Then all that needs to be done is invoke only simple RPCs with the backoff

Not in the client, for the reasons mentioned above.  We just need timeout to be implemented in this PR.

> I also would not recommend going back to a non-global backoff, since I do not see a value in that.

Top level retry is provided by the actor implementation.  That gives the most flexibility.  And there are plans to add additional backoff mechanisms and failure detection based on run times, etc.  And then it is needed only in one place.  If users need additional control, they can do it in their actors.

File: src/frequenz/client/electricity_trading/_client.py [hunk 5]
```diff
@@ -91,6 +92,31 @@ def validate_decimal_places(value: Decimal, decimal_places: int, name: str) -> N
...
+    Returns:
+        The result of the gRPC call.
+
+    Raises:
+        asyncio.TimeoutError: If the call exceeds the timeout.
+    """
+    try:
+        return await asyncio.wait_for(call(*args, **kwargs), timeout=timeout)
+    except asyncio.TimeoutError:
+        _logger.error("Timeout while calling %s", call)
```
Comment (shsms) [inline src/frequenz/client/electricity_trading/_client.py]:
```suggestion
        _logger.exception("Timeout while calling %s", call)

File: src/frequenz/client/electricity_trading/_client.py [hunk 6]
```diff
@@ -91,6 +92,31 @@ def validate_decimal_places(value: Decimal, decimal_places: int, name: str) -> N
         ) from exc
 
 
+async def grpc_call_with_timeout(
+    call: Callable[..., Awaitable[Any]], *args: Any, timeout: float = 300, **kwargs: Any
```
Comment (shsms) [inline src/frequenz/client/electricity_trading/_client.py]: I think this should take a `timedelta | None` and the default value should be `None` here and in the other calls.  Otherwise it wouldn't be intuitive and not what you'd expect from a thin wrapper.

Comment (ghost) [reply src/frequenz/client/electricity_trading/_client.py]: I agree about the `timedelta` but I don't think that the timeout should be `None` by default. Else we risk having the API stuck again (without giving any errors) like we did last week

Comment (shsms) [reply src/frequenz/client/electricity_trading/_client.py]: Well, the actor would set it when making calls to the client.  The actor can get the value to use from the config file.

The 300s comes from specific issues we noticed in the server and what we assume the actor needs.  We can't say that 300 is the right value in the long term.  It could become different even as soon as we go to production.

**Comment by tiyash-basu-frequenz** [comment]
> > Then all that needs to be done is invoke only simple RPCs with the backoff
> 
> Not in the client, for the reasons mentioned above. We just need timeout to be implemented in this PR.
> 
> > I also would not recommend going back to a non-global backoff, since I do not see a value in that.
> 
> Top level retry is provided by the actor implementation. That gives the most flexibility. And there are plans to add additional backoff mechanisms and failure detection based on run times, etc. And then it is needed only in one place. If users need additional control, they can do it in their actors.

So if I understand it right, you specifically are suggesting having per-RPC backoffs in this client? If yes, how would that help?

Nice that a global backoff module exists in the actor library!

**Comment by shsms** [comment]
> you specifically are suggesting having per-RPC backoffs in this client?

No, I'm saying we shouldn't add automatic retry in the client for non-streaming RPCs.  There is no question of backoff if there's no retry.

I'm only suggesting that we have timeout for such RPCs.

File: src/frequenz/client/electricity_trading/_client.py (hunk 6 above)
Comment (ghost) [reply src/frequenz/client/electricity_trading/_client.py]: No the 300s was arbitrary just to have _some_ timeout in the API requests. And the issues didn't come from the actor's side but from the API side when it disconnected (without throwing any errors). I don't mind changing the 300s to another value, but I'd feel more comfortable having a timeout by default.

Comment (shsms) [reply src/frequenz/client/electricity_trading/_client.py]: > No the 300s was arbitrary just to have some timeout in the API requests.

This is just the thing. Defaults have to make sense, and we can't find a value that does, especially when we expect external parties to use it. Maybe we shouldn't add this here at all and keep it only in the actors? If, at some point, we get an SDK-like layer between the client and the actor, these problems would become much easier.

Maybe @llucax has an opinion?  Because we try to be consistent with all the API clients.

In the worst case, we go ahead with 300s as the default if it works for most cases.

For example, the default receiver buffer size of 50 is a bit arbitrary and has bitten me sometimes. It didn't occur to me for a long time that it was overflowing and needed a bigger buffer. But still, having 50 as a default has been very useful in most places. The situation here is very different because `None` as buffer size would have meant no buffer, whereas here `None` as timeout means infinite timeout.

Comment (ghost) [reply src/frequenz/client/electricity_trading/_client.py]: Yea ok I get your point. I can then set the default to None for now then & just make sure we call the functions with timeouts in the actors

File: RELEASE_NOTES.md [hunk 7]
```diff
@@ -14,6 +14,7 @@
 * Extra validation check to ensure the quantity and price are within the allowed bounds.
 * Add more edge cases to the integration tests.
 * Add idiomatic string representations for `Power` and `Price` classes.
+* Add a timeout in the gRPC function calls
```
Comment (shsms) [inline RELEASE_NOTES.md]:
```suggestion
* Add support for timeouts in the gRPC function calls

**Review by shsms — DISMISSED** [review]
Just an optional comment.

File: RELEASE_NOTES.md (hunk 7 above)
Comment (ghost) [reply RELEASE_NOTES.md]: Done

**Review by shsms — APPROVED** [review]
